From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Fri, 7 Nov 2025 01:20:49 +0900
Subject: [PATCH] Skip distanceToSqr call in ServerEntity sendChanges

Project: SparklyPaper (https://github.com/SparklyPower/SparklyPaper)
Author: MrPowerGamerBR <git@mrpowergamerbr.com>
License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)

Description:
* This patch is based on SparklyPaper's "Skip "distanceToSqr" call in "ServerEntity#sendChanges" if the delta movement hasn't changed" patch.

The "distanceToSqr" call is a bit expensive, so avoiding it is pretty nice, around ~15% calls are skipped with this check

We could also check if the x,y,z coordinates are equal, but for now, let's just keep the identity check, which also helps us since Minecraft's code does reuse the original delta movement Vec3 object

diff --git a/net/minecraft/server/level/ServerEntity.java b/net/minecraft/server/level/ServerEntity.java
index e96d4dee14c05f2fa329bfb1588ec795d4e3d730..6f97359eae8f6fb645085c904b63e3ebc8d269d5 100644
--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -209,6 +209,7 @@ public class ServerEntity {
 
                 if (this.entity.hasImpulse || this.trackDelta || this.entity instanceof LivingEntity && ((LivingEntity)this.entity).isFallFlying()) {
                     Vec3 deltaMovement = this.entity.getDeltaMovement();
+                    if (deltaMovement != this.lastSentMovement) { // Plazma start - Skip distanceToSqr call in ServerEntity sendChanges
                     double d = deltaMovement.distanceToSqr(this.lastSentMovement);
                     if (d > 1.0E-7 || d > 0.0 && deltaMovement.lengthSqr() == 0.0) {
                         this.lastSentMovement = deltaMovement;
@@ -226,6 +227,7 @@ public class ServerEntity {
                             this.broadcast.accept(new ClientboundSetEntityMotionPacket(this.entity.getId(), this.lastSentMovement));
                         }
                     }
+                    } // Plazma end - Skip distanceToSqr call in ServerEntity sendChanges
                 }
 
                 if (packet != null) {
